engine/crates/grafbase-engine/src/resolver_utils/list.rs

```diff
@@ -1,129 +1,201 @@
-                                        .await
-                                        .insert_node(ResponsePrimitive::new(result.into())),
-                                ))
-                            }
-                            _ => resolve_container(&ctx_idx, ty, None, Some(item))
-                                .await
-                                .map(Option::Some)
-                                .map_err(|err| ctx_idx.set_error_path(err)),
-                        }
+        let mut children = vec![];
+        for result in futures_util::future::join_all(futures).await {
+            // Now we need to handle error propagation and validate the nullability
+            // of each of the list items
+            match result {
+                Ok(id) if list_kind.inner_nullablity() == ListInner::NonNullable => {
+                    let found_null = match ctx.response_graph.read().await.get_node(id) {
+                        Some(QueryResponseNode::Primitive(value)) if value.is_null() => true,
+                        None => true,
+                        _ => false,
                     };
-                    futures_util::pin_mut!(resolve_fut);
-                    extensions
-                        .resolve(resolve_info, &mut resolve_fut)
-                        .await
-                        .map(|value| value.expect("You definitely encountered a bug!"))
-                }
-            });
-        }
-        let node = ResponseList::with_children(futures_util::future::try_join_all(futures).await?);
-
-        Ok(ctx.response_graph.write().await.insert_node(node))
-    } else {
-        let mut futures = Vec::with_capacity(values.len());
-        for (idx, item) in values.into_iter().enumerate() {
-            let ctx_idx = ctx.with_index(idx, Some(&ctx.item.node));
-            futures.push(async move {
-                match ty {
-                    MetaType::Scalar { .. } | MetaType::Enum { .. } => {
-                        let result = Value::try_from(item.take()).map_err(|err| {
-                            ctx_idx.set_error_path(ServerError::new(format!("{err:?}"), Some(field.pos)))
-                        })?;
-
-                        Ok(ctx_idx
-                            .response_graph
-                            .write()
-                            .await
-                            .insert_node(ResponsePrimitive::new(result.into())))
+                    if found_null {
+                        ctx.add_error(ServerError::new(
+                            format!(
+                                "An error occurred while fetching `{}`, a non-nullable value was expected but nm value was found.",
+                                field.node.name.node
+                            ),
+                            Some(ctx.item.pos),
+                        ));
+                        children.push(ctx.response_graph.write().await.insert_node(CompactValue::Null));
```

Comment:
shouldn't you rather replace the node instead of adding a new one? The goal is to have a compacter representation with `CompactValue::Null` no?

Reply:
Not at all, might just be worth a comment. A bit surprising to see you adding a null when you already have a null value. So to be honest without taking a deeper look, I can't explain why.^^

Reply:
Actually never mind I'll just do it

Reply:
Also ignore anything I said above: I didn't look at the context of this line, thought we were talking about another bit of code